linguicator_predictor/service.py

```diff
@@ -0,0 +1,158 @@
+    :type  request_id: int
+
+    :param request_data: this was added to make all request handlers receive
+                         the same amount of arguments, but it is not necessary
+                         for the function to
+                         work properly
+    :type  request_data: any
+
+    :returns: the unaltered user text
+    :rtype: str
+    """
+    prediction = predict(user_text)
+    await websocket.send(json.dumps({
+        'id': request_id,
+        'code': 200,
+        'data': prediction
+    }))
+    return user_text
+
+
+def predict(user_text):
+    """
+    I make a prediction of how a text could continue.
+
+    :param user_text: the text I will make the prediction on
+    :type  user_text: str
+
+    :returns: the predicted text
+    :rtype: str
+    """
+    prediction = linguicator_predictor.model.predict(user_text,
+                                                     len(user_text) + 30)
+
+    # remove the original text from the prediction
+    prediction = prediction[len(user_text):]
+
+    # remove everything after last ' ' and return
+    return ' '.join(prediction.split()[:-1])
+
+
+async def handle_feed_request(websocket, user_text, request_id, new_text):
+    """
+    I handle a request add more text to the user typed text and make a
+    prediction if necessary.
+
+    :param websocket: websocket object representing the connection
+    :type  websocket: websockets.server.WebSocketServerProtocol
+
+    :param user_text: all the text inputed by the user so far
+    :type  user_text: str
+
+    :param request_id: the id of the request being handled
+    :type  request_id: int
+
+    :param new_text: the new piece of text inputed by the user
+    :type  new_text: str
+
+    :returns: the updated user text
+    :rtype: str
+    """
+    updated_user_text = user_text + new_text
+    if should_predict(updated_user_text):
+        await send_text_prediction(websocket, updated_user_text, request_id)
+    else:
+        await websocket.send(json.dumps({'id': request_id, 'code': 200}),
+                             ensure_ascii = False)
+    return updated_user_text
+
+
+async def send_text_prediction(websocket, user_text, request_id):
+    """
+    I send a prediction to the user.
+
+    :param websocket: websocket object representing the connection
+    :type  websocket: websockets.server.WebSocketServerProtocol
+
+    :param user_text: the text used to make the prediction
+    :type  user_text: str
+
+    :param request_id: the id of the request being handled
+    :type  request_id: int
+
+    :returns: nothing
+    :rtype: None
+    """
+    predicted_text = predict(user_text)
+    response_data = {
+        'id': request_id,
+        'code': 200,
+        'data': predicted_text
+    }
+    logging.info('sending data through websocket connection')
+    logging.debug(f'sending: {response_data}')
+    await websocket.send(json.dumps(response_data, ensure_ascii = False))
+
+
+def should_predict(text):
```

Comment:
If we change how the protocol works, deciding on which we should predict or not is not necessary, once every request from the front-end implies in a prediction.